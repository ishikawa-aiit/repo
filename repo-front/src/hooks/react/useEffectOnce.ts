import { useEffect, useRef, useState } from 'react'

/**
 * 依存関係がゼロの時、useEffect（）の代わりに使用するカスタムフック
 *
 * @remarks
 * このカスタムフックはReact 18の仕様変更のため、作成しています（React 18では依存関係がゼロの時、2回レンダリングされる）
 * https://github.com/vercel/next.js/issues/35822
 */
export const useEffectOnce = (effect: () => void | (() => void)) => {
  const destroyFunc = useRef<void | (() => void)>()
  const effectCalled = useRef(false)
  const renderAfterCalled = useRef(false)
  const [_val, setVal] = useState<number>(0)

  if (effectCalled.current) {
    renderAfterCalled.current = true
  }

  useEffect(() => {
    if (!effectCalled.current) {
      destroyFunc.current = effect()
      effectCalled.current = true
    }

    setVal((val) => {
      return val + 1
    })

    return () => {
      if (!renderAfterCalled.current) {
        return
      }
      if (destroyFunc.current) {
        destroyFunc.current()
      }
    }
    // マウント時に1回だけ実行するため
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])
}
