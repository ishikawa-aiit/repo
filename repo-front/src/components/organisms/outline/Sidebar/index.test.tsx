import { render, screen, within } from '@testing-library/react'

import { Sidebar } from '.'

jest.mock('next/router', () => {
  return {
    useRouter: () => {
      return {
        pathname: '/weekly-reports/create',
      }
    },
  }
})

describe('Sidebar', () => {
  it('正常にレンダリングされている', () => {
    const { container } = render(<Sidebar />)

    expect(container).toMatchSnapshot()
  })

  describe('ルーターから返ってくるパス名(pathname)とmenuItemsおよびbottomMenuItems内のlink', () => {
    it('一致する場合、カレントページとなる', () => {
      const { container } = render(<Sidebar />)
      const { getByTestId } = within(screen.getByTestId('Sidebar'))
      const { getAllByTestId } = within(screen.getByTestId('Sidebar'))

      expect(getByTestId('SidebarCurrentMenuItem')).toHaveTextContent('週報作成')

      const sidebarMenuItems = getAllByTestId('SidebarMenuItem')
      sidebarMenuItems.forEach((sidebarMenuItem) => {
        expect(sidebarMenuItem).not.toHaveTextContent('週報作成')
      })
      expect(container).toMatchSnapshot()
    })
  })
})
