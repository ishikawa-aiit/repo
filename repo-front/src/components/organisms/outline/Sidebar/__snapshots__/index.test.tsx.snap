// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar ルーターから返ってくるパス名(pathname)とmenuItemsおよびbottomMenuItems内のlink 一致する場合、カレントページとなる 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: -webkit-sticky;
  position: sticky;
  background: primary.500;
  width: 240px;
  top: 68px;
  height: calc(100vh - 68px);
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  color: text.white;
  font-size: sm;
  gap: 4px;
  font-weight: bold;
  -webkit-padding-start: 6px;
  padding-inline-start: 6px;
  -webkit-padding-end: 6px;
  padding-inline-end: 6px;
  padding-top: 4px;
  padding-bottom: 4px;
  -webkit-text-decoration: none;
  text-decoration: none;
  transition-duration: .25s;
  cursor: pointer;
}

.emotion-2[aria-current=page] {
  background-color: primary.600;
}

.emotion-3 {
  width: 5px;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  color: text.white;
  font-size: sm;
  gap: 4px;
  font-weight: bold;
  -webkit-padding-start: 6px;
  padding-inline-start: 6px;
  -webkit-padding-end: 6px;
  padding-inline-end: 6px;
  padding-top: 4px;
  padding-bottom: 4px;
  -webkit-text-decoration: none;
  text-decoration: none;
  transition-duration: .25s;
  cursor: pointer;
}

.emotion-4[aria-current=page] {
  background-color: primary.600;
}

.emotion-4:hover,
.emotion-4[data-hover] {
  background: primary.400;
}

.emotion-8 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  justify-self: stretch;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
}

<div>
  <aside
    class="emotion-0"
    data-cy="Sidebar"
    data-testid="Sidebar"
  >
    <div
      class="emotion-1"
    >
      <a
        aria-current="page"
        class="emotion-2"
        data-cy="SidebarCurrentMenuItem"
        data-testid="SidebarCurrentMenuItem"
      >
        <div
          class="emotion-3"
        >
          <svg
            aria-hidden="true"
            fill="none"
            stroke="currentColor"
            stroke-width="2"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11 5H6a2 2 0 00-2 2v11a2 2 0 002 2h11a2 2 0 002-2v-5m-1.414-9.414a2 2 0 112.828 2.828L11.828 15H9v-2.828l8.586-8.586z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
        週報作成
      </a>
      <a
        class="emotion-4"
        data-cy="SidebarMenuItem"
        data-testid="SidebarMenuItem"
      >
        <div
          class="emotion-3"
        >
          <svg
            aria-hidden="true"
            fill="none"
            stroke="currentColor"
            stroke-width="2"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4 6h16M4 12h16M4 18h7"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
        週報一覧
      </a>
      <a
        class="emotion-4"
        data-cy="SidebarMenuItem"
        data-testid="SidebarMenuItem"
      >
        <div
          class="emotion-3"
        >
          <svg
            aria-hidden="true"
            fill="none"
            stroke="currentColor"
            stroke-width="2"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4 5a1 1 0 011-1h14a1 1 0 011 1v2a1 1 0 01-1 1H5a1 1 0 01-1-1V5zM4 13a1 1 0 011-1h6a1 1 0 011 1v6a1 1 0 01-1 1H5a1 1 0 01-1-1v-6zM16 13a1 1 0 011-1h2a1 1 0 011 1v6a1 1 0 01-1 1h-2a1 1 0 01-1-1v-6z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
        テンプレート作成・編集
      </a>
    </div>
    <div
      class="emotion-8"
    />
    <div
      class="emotion-1"
    >
      <a
        class="emotion-4"
        data-cy="SidebarMenuItem"
        data-testid="SidebarMenuItem"
      >
        <div
          class="emotion-3"
        >
          <svg
            aria-hidden="true"
            fill="none"
            stroke="currentColor"
            stroke-width="2"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10.325 4.317c.426-1.756 2.924-1.756 3.35 0a1.724 1.724 0 002.573 1.066c1.543-.94 3.31.826 2.37 2.37a1.724 1.724 0 001.065 2.572c1.756.426 1.756 2.924 0 3.35a1.724 1.724 0 00-1.066 2.573c.94 1.543-.826 3.31-2.37 2.37a1.724 1.724 0 00-2.572 1.065c-.426 1.756-2.924 1.756-3.35 0a1.724 1.724 0 00-2.573-1.066c-1.543.94-3.31-.826-2.37-2.37a1.724 1.724 0 00-1.065-2.572c-1.756-.426-1.756-2.924 0-3.35a1.724 1.724 0 001.066-2.573c-.94-1.543.826-3.31 2.37-2.37.996.608 2.296.07 2.572-1.065z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
            <path
              d="M15 12a3 3 0 11-6 0 3 3 0 016 0z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
        設定
      </a>
    </div>
  </aside>
</div>
`;

exports[`Sidebar 正常にレンダリングされている 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: -webkit-sticky;
  position: sticky;
  background: primary.500;
  width: 240px;
  top: 68px;
  height: calc(100vh - 68px);
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  color: text.white;
  font-size: sm;
  gap: 4px;
  font-weight: bold;
  -webkit-padding-start: 6px;
  padding-inline-start: 6px;
  -webkit-padding-end: 6px;
  padding-inline-end: 6px;
  padding-top: 4px;
  padding-bottom: 4px;
  -webkit-text-decoration: none;
  text-decoration: none;
  transition-duration: .25s;
  cursor: pointer;
}

.emotion-2[aria-current=page] {
  background-color: primary.600;
}

.emotion-3 {
  width: 5px;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  color: text.white;
  font-size: sm;
  gap: 4px;
  font-weight: bold;
  -webkit-padding-start: 6px;
  padding-inline-start: 6px;
  -webkit-padding-end: 6px;
  padding-inline-end: 6px;
  padding-top: 4px;
  padding-bottom: 4px;
  -webkit-text-decoration: none;
  text-decoration: none;
  transition-duration: .25s;
  cursor: pointer;
}

.emotion-4[aria-current=page] {
  background-color: primary.600;
}

.emotion-4:hover,
.emotion-4[data-hover] {
  background: primary.400;
}

.emotion-8 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  justify-self: stretch;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
}

<div>
  <aside
    class="emotion-0"
    data-cy="Sidebar"
    data-testid="Sidebar"
  >
    <div
      class="emotion-1"
    >
      <a
        aria-current="page"
        class="emotion-2"
        data-cy="SidebarCurrentMenuItem"
        data-testid="SidebarCurrentMenuItem"
      >
        <div
          class="emotion-3"
        >
          <svg
            aria-hidden="true"
            fill="none"
            stroke="currentColor"
            stroke-width="2"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11 5H6a2 2 0 00-2 2v11a2 2 0 002 2h11a2 2 0 002-2v-5m-1.414-9.414a2 2 0 112.828 2.828L11.828 15H9v-2.828l8.586-8.586z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
        週報作成
      </a>
      <a
        class="emotion-4"
        data-cy="SidebarMenuItem"
        data-testid="SidebarMenuItem"
      >
        <div
          class="emotion-3"
        >
          <svg
            aria-hidden="true"
            fill="none"
            stroke="currentColor"
            stroke-width="2"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4 6h16M4 12h16M4 18h7"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
        週報一覧
      </a>
      <a
        class="emotion-4"
        data-cy="SidebarMenuItem"
        data-testid="SidebarMenuItem"
      >
        <div
          class="emotion-3"
        >
          <svg
            aria-hidden="true"
            fill="none"
            stroke="currentColor"
            stroke-width="2"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4 5a1 1 0 011-1h14a1 1 0 011 1v2a1 1 0 01-1 1H5a1 1 0 01-1-1V5zM4 13a1 1 0 011-1h6a1 1 0 011 1v6a1 1 0 01-1 1H5a1 1 0 01-1-1v-6zM16 13a1 1 0 011-1h2a1 1 0 011 1v6a1 1 0 01-1 1h-2a1 1 0 01-1-1v-6z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
        テンプレート作成・編集
      </a>
    </div>
    <div
      class="emotion-8"
    />
    <div
      class="emotion-1"
    >
      <a
        class="emotion-4"
        data-cy="SidebarMenuItem"
        data-testid="SidebarMenuItem"
      >
        <div
          class="emotion-3"
        >
          <svg
            aria-hidden="true"
            fill="none"
            stroke="currentColor"
            stroke-width="2"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10.325 4.317c.426-1.756 2.924-1.756 3.35 0a1.724 1.724 0 002.573 1.066c1.543-.94 3.31.826 2.37 2.37a1.724 1.724 0 001.065 2.572c1.756.426 1.756 2.924 0 3.35a1.724 1.724 0 00-1.066 2.573c.94 1.543-.826 3.31-2.37 2.37a1.724 1.724 0 00-2.572 1.065c-.426 1.756-2.924 1.756-3.35 0a1.724 1.724 0 00-2.573-1.066c-1.543.94-3.31-.826-2.37-2.37a1.724 1.724 0 00-1.065-2.572c-1.756-.426-1.756-2.924 0-3.35a1.724 1.724 0 001.066-2.573c-.94-1.543.826-3.31 2.37-2.37.996.608 2.296.07 2.572-1.065z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
            <path
              d="M15 12a3 3 0 11-6 0 3 3 0 016 0z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
        設定
      </a>
    </div>
  </aside>
</div>
`;
